### PR TITLE
Hmac auth body validation

### LINKED ISSUE #3345: hmac-auth with body validation enabled fails to validate if no digest header is present
### Summary

When you enable body validation on the hmac-auth plugin, the plugin will accept a request without a digest header.

### Steps To Reproduce

1. setup the consumer and the plugin
```
curl -X DELETE http://localhost:8001/consumers/alice

curl -i -X POST http://localhost:8001/consumers/ \
      -d username=alice

curl -i -X POST http://localhost:8001/consumers/alice/hmac-auth \
      -d username=alice123 \
      -d secret=secret

curl -X DELETE http://localhost:8001/apis/example-service

curl -i -X POST http://localhost:8001/apis \
     -d name=example-service   \
     -d hosts=hmac.com \
     -d upstream_url=https://httpbin.org/anything


curl -i -X POST http://localhost:8001/apis/example-service/plugins \
      -d name=hmac-auth \
      -d config.enforce_headers=date,host \
      -d config.validate_request_body=true \
      -d config.algorithms=hmac-sha256

```
2.  this works
```
export BODY="a small body"
export DIGEST="SHA-256=$(echo -n "$BODY" | openssl sha -binary -sha256 | base64)"

export DATE=$(date -u +%a,\ %d\ %b\ %Y\ %H:%M:%S\ GMT)
export HMAC_MESSAGE=$(echo "host: hmac.com" ; echo -n "date: $DATE")
export SIGNATURE=$( echo -n "$HMAC_MESSAGE" | openssl sha -binary -sha256 -hmac secret | base64)

curl -i -X GET http://localhost:8000 \
      -H "host: hmac.com" \
      -H "date: $DATE" \
      -H "digest: $DIGEST" \
      -H "authorization: hmac username=\"alice123\", algorithm=\"hmac-sha256\", headers=\"host date\", signature=\"$SIGNATURE\"" \
      -d "$BODY"
```
3. this works but should fail
```
curl -i -X GET http://localhost:8000 \
      -H "host: hmac.com" \
      -H "date: $DATE" \
      -H "authorization: hmac username=\"alice123\", algorithm=\"hmac-sha256\", headers=\"host date\", signature=\"$SIGNATURE\"" \
      -d "$BODY"
```
### Additional Details & Logs

- Kong version (`$ kong version`)
0.11.2 and upward

### LINKED ISSUE #3346: When you enable body validation on the hmac-auth plugin, the plugin will fail on requests without a body
### Summary

When you enable body validation on the hmac-auth plugin, the plugin will fail on requests without a body

### Steps To Reproduce

1. setup the consumer, api and plugin
```
curl -X DELETE http://localhost:8001/consumers/alice

curl -i -X POST http://localhost:8001/consumers/ \
      -d username=alice

curl -i -X POST http://localhost:8001/consumers/alice/hmac-auth \
      -d username=alice123 \
      -d secret=secret

curl -X DELETE http://localhost:8001/apis/example-service

curl -i -X POST http://localhost:8001/apis \
     -d name=example-service   \
     -d hosts=hmac.com \
     -d upstream_url=https://httpbin.org/anything


curl -i -X POST http://localhost:8001/apis/example-service/plugins \
      -d name=hmac-auth \
      -d config.enforce_headers=date,host \
      -d config.validate_request_body=true \
      -d config.algorithms=hmac-sha256
```

2. this works
```
export BODY="a small body"
export DATE=$(date -u +%a,\ %d\ %b\ %Y\ %H:%M:%S\ GMT)
export HMAC_MESSAGE=$(echo "host: hmac.com" ; echo -n "date: $DATE")
export SIGNATURE=$( echo -n "$HMAC_MESSAGE" | openssl sha -binary -sha256 -hmac secret | base64)
export DIGEST=$(echo -n "$BODY" | openssl sha -binary -sha256 | base64)

curl -i -X GET http://localhost:8000 \
      -H "host: hmac.com" \
      -H "date: $DATE" \
      -H "digest: SHA-256=$DIGEST" \
      -H "authorization: hmac username=\"alice123\", algorithm=\"hmac-sha256\", headers=\"host date\", signature=\"$SIGNATURE\"" \
      -d "$BODY"
```
3. this does not work
```
export SIGNATURE=$( openssl sha -binary -sha256   < /dev/null | base64)
curl -i -X GET http://localhost:8000 \
      -H "host: hmac.com" \
      -H "digest: SHA-256=$DIGEST" \
      -H "date: $DATE" \
      -H "authorization: hmac username=\"alice123\", algorithm=\"hmac-sha256\", headers=\"host date\", signature=\"$SIGNATURE\"" 
```
### Additional Details & Logs
- Kong version (`$ kong version`)
0.11.2 and upwards.

### PR DESCRIPTION
**Pull request by mvanholsteijn**
### Summary
Changes the hmac-auth request validation logic to pass if:

- there is no `Digest` and no body
- there is `Digest` for an empty body  and no body.
- there is a `Digest` for the body and a body.

This is to make sure that we can put 'digest' in the enforce_headers for all requests.

## Full changelog

* added tests for scenarios
* added logic in validate_body()

### Issues resolved

Fix #3345 
Fix #3346

### CODE REVIEW

**Comment by shashiranjan84** [comment]
@mvanholsteijn thanks for PR, would you please rebase before we start reviewing ?

**Comment by mvanholsteijn** [comment]
@shashiranjan84 Do you have any idea on when this PR might be reviewed?

File: kong/plugins/hmac-auth/access.lua [hunk 1]
```diff
@@ -206,24 +206,25 @@ local function validate_clock_skew(headers, date_header_name, allowed_clock_skew
...
+  if not digest_received then
+    -- if there is no digest and no body, it is ok
+    return not body
+  end
+
   if not body then
-    return false
+    -- if no body, calculate sha-256 over 0 bytes
+    body = ''
   end
```
Comment (shashiranjan84) [inline kong/plugins/hmac-auth/access.lua]: you can drop this check and just do 
```
sha256:update(body or '')
```

**Comment by shashiranjan84** [comment]
@mvanholsteijn would you please also update the doc, https://github.com/Kong/getkong.org/blob/master/app/plugins/hmac-authentication.md

**Comment by mvanholsteijn** [comment]
@shashiranjan84 I added the documentation to getkong.org...

**Comment by thibaultcha** [comment]
@mvanholsteijn Thank you for the patch! If you haven't done so yet, be sure to grab your [Contributor T-shirt](https://github.com/Kong/kong/blob/master/CONTRIBUTING.md#contributor-t-shirt) :)

**Pull request merged**